src/k8s/pkg/k8sd/setup/certificates.go

```diff
@@ -11,54 +11,90 @@ import (
 )
 
 // ensureFile creates fname with the specified contents, mode and owner bits.
-// ensureFile will delete the file if contents is an empty string.
-func ensureFile(fname string, contents string, uid, gid int, mode fs.FileMode) error {
+// It will delete the file if contents parameter is an empty string. Trying to ensure a inexistent file
+// with an empty contents parameter does not result in an error.
+// It returns true if any of these is true: the file's content changed, it was created or it was deleted.
+// It also returns any error that occured.
+func ensureFile(fname string, contents string, uid, gid int, mode fs.FileMode) (bool, error) {
 	if contents == "" {
-		if err := os.Remove(fname); err != nil && !os.IsNotExist(err) {
-			return fmt.Errorf("failed to delete: %w", err)
+		if err := os.Remove(fname); err != nil {
+			if !os.IsNotExist(err) {
+				// File exists but failed to delete.
+				return false, fmt.Errorf("failed to delete: %w", err)
+			}
+			// File does not exist, nothing to do.
+			return false, nil
 		}
-		return nil
+
+		// File was deleted.
+		return true, nil
+	}
+
+	origContent, err := os.ReadFile(fname)
+	if err != nil && !os.IsNotExist(err) {
+		// File exists but failed to read.
+		return false, fmt.Errorf("failed to read: %w", err)
 	}
 
-	if err := os.WriteFile(fname, []byte(contents), mode); err != nil {
-		return fmt.Errorf("failed to write: %w", err)
+	var contentChanged bool
+
+	if contents != string(origContent) {
+		if err := os.WriteFile(fname, []byte(contents), mode); err != nil {
+			return false, fmt.Errorf("failed to write: %w", err)
+		} else {
+			contentChanged = true
+		}
```

Comment:
no need for an else block here
```suggestion
		}
		contentChanged = true
```